### PR TITLE
[Part 1/n] Rename Dataset => Datastream in top level files and pydocs

### PR DESCRIPTION
**Pull request by ericl**
## Context

As we are enabling streaming datasets by default in Ray 2.4, it's time to revisit the Datasets naming conundrum.

A few things we know for sure:
- "Ray Dataset" puts us in the same space as DataFrame / ETL libraries like Spark, which is a common confusion.
- Users are a little confused about "Ray Data" vs "Ray Dataset", however, this isn't a big deal--- more like a nit.

Goals:
- Avoid unnecessary churn and confusion
- Clearly position Ray Data as a specialized library for ML training ingest and batch inference
- Technical accuracy of the name (i.e., don't over-correct on positioning and come up with technically misleading component names).

## Proposal

1. Rename `Dataset` to `Datastream` (and secondary renamings like DatasetContext, DatasetIterator).
2. When a `Datastream` is cached or otherwise fully materialized in memory (e.g., created via from_arrow_refs), it is instead created as a subclass of Datastream: `MaterializedDatastream(Datastream)`.

This seems to solve the positioning challenges, while preserving the technical accuracy of the class names (i.e., it's never called a "stream" when there's no streaming involved).

## This PR

This PR starts by renaming the top-level classes and updating pydoc in the top-level files. In future PRs:
- [ ] Rename secondary APIs like DatasetContext, GroupedDataset, etc.
- [ ] Update internal references to Datastream type
- [ ] Rename dataset.py to datastream.py
- [ ] Update docs and example code

A few examples:
```
>>> ray.data.from_items([1,2,3,4,5])
MaterializedDatastream(num_blocks=5, num_rows=5, schema=<class 'int'>)

>>> ds = ray.data.range_tensor(10000)
Datastream(
   num_blocks=16,
   num_rows=10000,
   schema={__value__: ArrowTensorType(shape=(1,), dtype=int64)}
)

>>> ds.split(2)
[MaterializedDatastream(
   num_blocks=8,
   num_rows=5000,
   schema={__value__: ArrowTensorType(shape=(1,), dtype=int64)}
), MaterializedDatastream(
   num_blocks=8,
   num_rows=5000,
   schema={__value__: ArrowTensorType(shape=(1,), dtype=int64)}
)]

>>> ds.cache()
MaterializedDatastream(
   num_blocks=16,
   num_rows=10000,
   schema={__value__: ArrowTensorType(shape=(1,), dtype=int64)}
)

```

### CODE REVIEW

**Comment by c21** [comment]
A little more research on other system, curious is the Datastream terminology here coming from Flink ([doc](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/overview/))?

Have you considered a name of `CachedDataStream` (this is same as [Flink](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/operators/overview/#datastream-rarr-cacheddatastream)) or `DataStream(cached=True)`, compared to `InMemoryData`?

**Comment by ericl** [comment]
I think a close analogy is Java 8 Streams: https://stackify.com/streams-guide-java-8/

In their API, Streams are a way to manipulate Collections. The Stream class represents a series of transformations on some base collection, similar to how in Ray Data the Datastream class represents a series of operations executed on some base input (either Datasource or InMemoryData). A Stream also be collected back to a Collection like a Datastream can be materialized into an InMemoryTable.

**Comment by stephanie-wang** [comment]
> A little more research on other system, curious is the Datastream terminology here coming from Flink ([doc](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/overview/))?
> 
> Have you considered a name of `CachedDataStream` (this is same as [Flink](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/operators/overview/#datastream-rarr-cacheddatastream)) or `DataStream(cached=True)` (this is my first reaction), compared to `InMemoryData`?

I also prefer the `CachedDataStream` naming here over `InMemoryData`.

**Comment by ericl** [comment]
Can you guys clarify why you like CachedDataStream? My mental model is this:

InMemoryData:
- from_items() -> InMemoryData   # This makes sense.
- cache() -> InMemoryData   # Not perfect but makes sense.
- split() -> List[InMemoryData]   # Not perfect but makes sense.

CachedDataStream:
- from_items() -> CachedDataStream   # This is odd.
- cache() -> CachedDataStream  # This makes sense.
- split() -> List[CachedDataStream]   # This is odd (implies the outputs are streaming, but they are not).

In other words, a more collection-oriented name seems to avoid sounding odd in all scenarios, even though it makes less sense for cache(). Btw, Flink's CachedDataStream is a lazily-cached stream, not a materialized data collection like we have here.

Another option is `MaterializedData[set]`, since "materialization" is a term I've seen users bring up organically, and it doesn't imply storage (e.g., "memory").

**Comment by c21** [comment]
btw a separate nit - can we name it `DataStream` instead of `Datastream`?

This would be same with rest of world (DataLoader, DataPipe, DataFrame, etc).

**Comment by jianoaix** [comment]
I'd +1 to `MaterializedData`, because 1) once it's executed, it's not quite a "stream" anymore; 2) `MaterializedDataset` seems to bring back the `Dataset` concept/confusion. Maybe the only slight downside is that "materialize" doesn't match the "cache" verb in API name.

**Comment by ericl** [comment]
@jianoaix I ended up calling it MaterializedDatastream in this PR, because it sounded better when writing the pydoc. MaterializedData / MaterializedDataset were harder to explain as return types.

**Comment by c21** [comment]
+1 to MaterializedDatastream, this sounds better than MaterializedData / MaterializedDataset.

**Review by jianoaix — APPROVED** [review]

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -1589,44 +1598,44 @@ def train_test_split(
...
+                    f"than the size of the datastream ({ds_length}). "
                     f"Got {test_size}."
                 )
-            return dataset.split_at_indices([dataset_length - test_size])
+            return ds.split_at_indices([ds_length - test_size])
 
     @ConsumptionAPI(pattern="Args:")
-    def union(self, *other: List["Dataset[T]"]) -> "Dataset[T]":
-        """Combine this dataset with others of the same type.
+    def union(self, *other: List["Datastream[T]"]) -> "MaterializedDatastream[T]":
```
Comment (c21) [inline python/ray/data/dataset.py:1613]: Can we keep the return value as `Datastream` instead of `MaterializedDatastream`? We will soon implement the union as a streaming physical operator (https://github.com/ray-project/ray/issues/32598) during Ray 2.5.

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -2126,22 +2135,22 @@ def sort(
                 descending=descending,
             )
             logical_plan = LogicalPlan(op)
-        return Dataset(plan, self._epoch, self._lazy, logical_plan)
+        return Datastream(plan, self._epoch, self._lazy, logical_plan)
 
-    def zip(self, other: "Dataset[U]") -> "Dataset[(T, U)]":
-        """Zip this dataset with the elements of another.
+    def zip(self, other: "Datastream[U]") -> "MaterializedDatastream[(T, U)]":
```
Comment (c21) [inline python/ray/data/dataset.py:2140]: Zip is lazy, so the return value should be `Datastream`.

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -2172,14 +2181,14 @@ def zip(self, other: "Dataset[U]") -> "Dataset[(T, U)]":
         if logical_plan is not None and other_logical_plan is not None:
             op = Zip(logical_plan.dag, other_logical_plan.dag)
             logical_plan = LogicalPlan(op)
-        return Dataset(plan, self._epoch, self._lazy, logical_plan)
+        return Datastream(plan, self._epoch, self._lazy, logical_plan)
 
     @ConsumptionAPI
-    def limit(self, limit: int) -> "Dataset[T]":
-        """Truncate the dataset to the first ``limit`` records.
+    def limit(self, limit: int) -> "MaterializedDatastream[T]":
```
Comment (c21) [inline python/ray/data/dataset.py:2187]: Can we keep the return value as Datastream instead of MaterializedDatastream? We will soon implement the union as a streaming physical operator (https://github.com/ray-project/ray/issues/32761) during Ray 2.5.

File: python/ray/data/dataset.py [hunk 4]
```diff
@@ -4055,52 +4062,53 @@ def __iter__(self):
...
 
         Returns:
-            A Dataset with all blocks fully materialized in memory.
+            A Datastream with all blocks fully materialized in memory.
         """
-        self._plan.execute(force_read=True)
-        return self
+        copy = Datastream.copy(self, _as=MaterializedDatastream)
+        copy._plan.execute(force_read=True)
+        return copy
```
Comment (c21) [inline python/ray/data/dataset.py:4105]: I feel returning a new object would make the semantic a little confusing. `ds.cache()` is having side effect on `ds`, but `ds` itself is still a `Datastream` with `is_cached() == True`. Feeling overloaded that we have two objects

* `Datastream` with `is_cached() == True` (this is quite weird)
* `MaterializedDatastream`

```
>>> ds = ray.data.range(10)
>>> output_ds = ds.cache()
>>> output_ds
MaterializedDatastream(num_blocks=10, num_rows=10, schema=<class 'int'>)
>>> ds.is_cached()
True
>>> ds
Datastream(num_blocks=10, num_rows=10, schema=<class 'int'>)
```

Comment (ericl) [reply python/ray/data/dataset.py:4105]: I'll try to fix it so there isn't a side effect. That was not intended originally.

Comment (ericl) [reply python/ray/data/dataset.py:4105]: Fixed by using a deep-copy.

File: python/ray/data/dataset.py [hunk 5]
```diff
@@ -261,21 +262,28 @@ def __init__(
...
-        return Dataset(dataset._plan, dataset._epoch, dataset._lazy)
+    def copy(
+        ds: "Datastream[T]", _deep_copy: bool = False, _as: Optional[type] = None
+    ) -> "Datastream[T]":
+        if not _as:
+            _as = Datastream
+        if _deep_copy:
+            return _as(ds._plan.copy(), ds._epoch, ds._lazy)
+        else:
+            return _as(ds._plan.deep_copy(), ds._epoch, ds._lazy)
```
Comment (c21) [inline python/ray/data/dataset.py:277]: shouldn't it be:

```
if _deep_copy:
    return _as(ds._plan.deep_copy(), ds._epoch, ds._lazy)
else:
    return _as(ds._plan.copy(), ds._epoch, ds._lazy)
```

Comment (ericl) [reply python/ray/data/dataset.py]: Fixed and added a test for it.

File: python/ray/data/dataset.py [hunk 6]
```diff
@@ -4047,52 +4056,53 @@ def __iter__(self):
...
+                lambda ds, read_stage=read_stage: Datastream(
                     ds._plan.with_stage(read_stage), ds._epoch, True
                 )
             )
         return pipe
 
-    @Deprecated(message="Use `Dataset.cache()` instead.")
-    def fully_executed(self) -> "Dataset[T]":
+    @Deprecated(message="Use `Datastream.cache()` instead.")
+    def fully_executed(self) -> "MaterializedDatastream[T]":
```
Comment (c21) [inline python/ray/data/dataset.py:4066]: after this PR, `ds.fully_executed()/cache()` will have behavior change, that `ds.fully_executed()/cache()` will trigger execution each time of call. Before this PR, only first call of `ds.fully_executed()/cache()` triggers execution.

Do we want to document the behavior change somewhere in doc?

**Review by c21 — APPROVED** [review]

File: python/ray/data/dataset.py (hunk 6 above)
Comment (ericl) [reply python/ray/data/dataset.py:4066]: I added a bit of explanation to the pydoc here, but I don't believe we need to further justify the previous behavior since that was never well defined / documented (in fact it seems like a bug).

Comment (c21) [reply python/ray/data/dataset.py:4066]: Ok, I was worried about people depending on the behavior. Cool, let's start with this.

Comment (ericl) [reply python/ray/data/dataset.py:4066]: It seemed like this did affect one existing code path--- one test with DatasetPipeline with an optimization flag disabled. Since this path is not documented, I opted to remove it. We could also roll back the behavior (have the mutating behavior) as an alternative, and delay changing this until we remove DatasetPipeline. However, I think the likelihood someone is depending on this flag is low.

Comment (c21) [reply python/ray/data/dataset.py:4057]: It's fine with me to change the behavior here.
